Add straight implementation of definitions

akscram opened the pull request:
The json_schema function got an additional positional argument which
allows to store and access references during composition of a scheme.

This change can break backward compatibility in case if the json_schema
function was used for schemas which contain TypeRef objects and which
are not considered as top level schemas.

akscram committed:
Add straight implementation of definitions

The json_schema function got an additional positional argument which
allows to store and access references during composition of a schema.

The refs argument could be None (which is set by default) or an instance
of the References class. If json_schema is called with refs=None and the
schema argument contains TypeRef objects, then the definitions item will
be added into the resulting schema with a value which will contain
schemas of TypeRef objects. If json_schema is called with an instance of
References, then the passed object will be used to store found
references during composition of a schema, hovewer the resulting schema
will not contain the definitions item and handling of definitions have
to be done explicetely.

This change can break backward compatibility in case if the json_schema
function was used for schemas which contain TypeRef objects and which
are not considered as top level schemas.

maximkulkin commented on tests/test_jsonschema.py:
```diff
@@ -316,3 +319,76 @@ def test_one_of_schema_with_mapping(self):
+        result = json_schema(TYPES['AType'], refs=refs)
+        definitions = refs.definitions()
+
+        assert bool(refs)
```
What this is for?

akscram replied:
To check that `refs` is not empty, it must not be for this particular case.

maximkulkin commented on tests/test_jsonschema.py:
```diff
@@ -316,3 +319,76 @@ def test_one_of_schema_with_mapping(self):
+            assert '$ref' in result['items']
+
+        TYPES = lr.TypeRegistry()
+        MyType = TYPES.add('MyType', lt.List(TYPES['MyType']))
```
This type is kind of unnatural. Can you change it to something like
```
lt.List(lt.OneOf([lt.String(), TYPES['MyType']]))
```

akscram replied:
Done.

maximkulkin commented on tests/test_jsonschema.py:
```diff
@@ -1,7 +1,10 @@
 from lollipop.utils import is_mapping
 from lollipop_jsonschema import json_schema
+from lollipop_jsonschema.compat import iteritems
+from lollipop_jsonschema.jsonschema import References
```
I think we need to make it public API to support cases when someone will want to accumulate definitions across multiple calls to json_schema.

akscram replied:
Done.

maximkulkin commented on tests/test_jsonschema.py:
```diff
@@ -316,3 +319,76 @@ def test_one_of_schema_with_mapping(self):
+            assert 'type' in result
+            assert 'array' == result['type']
+            assert 'items' in result
+            assert '$ref' in result['items']
```
'$ref' is not a mandatory attribute for an array.

akscram replied:
That is not about arrays, it is about how references work, e.g. each instance of TypeRef is represented as `{"$ref": <url>}`. So, here I am checking that `items` is a reference for this particular case.

akscram committed:
Exclude setuptools==36.2.0 from dependencies

The original problem is tracked in:
    https://github.com/pypa/setuptools/pull/1085
